### PR TITLE
[Feature] filter partition with join predicate when refresh mv

### PR DESCRIPTION
**Pull request by choury**
In the materializedView class, add a partitionExprMaps field to store expressions equivalent to the partition column of the materialized view, collected from join predicates. With this information, we can identify the related partition tables, which will undergo partition pruning during the refresh of the materialized view.
Regarding the union statement, by merging all sub-clauses, we can obtain partitionExprMaps.

## What type of PR is this:

- [ ] BugFix
- [x] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

Does this PR entail a change in behavior?

- [ ] Yes, this PR will result in a change in behavior.
- [x] No, this PR will not result in a change in behavior.

If yes, please specify the type of change:

- [ ] Interface/UI changes: syntax, type conversion, expression evaluation, display information
- [ ] Parameter changes: default values, similar parameters but with different default values
- [ ] Policy changes: use new policy to replace old one, functionality automatically enabled
- [ ] Feature removed
- [ ] Miscellaneous: upgrade & downgrade compatibility, etc.

## Checklist:

- [ ] I have added test cases for my bug fix or my new feature
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:

- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [ ] 3.2
  - [ ] 3.1
  - [ ] 3.0
  - [ ] 2.5

### CODE REVIEW

File: fe/fe-core/src/main/java/com/starrocks/catalog/MaterializedView.java [hunk 1]
```diff
@@ -1584,6 +1694,22 @@ public void gsonPostProcess() throws IOException {
...
+            for (Map.Entry<GsonUtils.ExpressionSerializedObject, GsonUtils.ExpressionSerializedObject> entry :
+                    serializedPartitionExprMaps.entrySet()) {
+                if (entry.getKey() != null && entry.getValue() != null) {
+                    partitionExprMaps.put(
+                            SqlParser.parseSqlToExpr(entry.getKey().expressionSql, SqlModeHelper.MODE_DEFAULT),
+                            (SlotRef) SqlParser.parseSqlToExpr(entry.getValue().expressionSql, SqlModeHelper.MODE_DEFAULT)
+                    );
+                }
+            }
+        } else if (!partitionRefTableExprs.isEmpty()) {
```
[the remark this answers is not in the document]
Comment (choury) [reply fe/fe-core/src/main/java/com/starrocks/catalog/MaterializedView.java:1708]: Yes, avoid empty `partitionRefTablExprs`

File: fe/fe-core/src/main/java/com/starrocks/sql/common/SyncPartitionUtils.java [hunk 2]
```diff
@@ -274,6 +269,54 @@ public static PartitionMapping mappingRange(Range<PartitionKey> baseRange, Strin
         return new PartitionMapping(truncLowerDateTime, truncUpperDateTime);
     }
 
+    // TODO: should support like to_date(ds) + 1 day ?
+    public static Range<PartitionKey> transferRange(Range<PartitionKey> baseRange, Expr partitionExpr)
```
[the remark this answers is not in the document]
Comment (choury) [reply fe/fe-core/src/main/java/com/starrocks/sql/common/SyncPartitionUtils.java:273]: I want to add support for more function like `to_date` later

File: fe/fe-core/src/main/java/com/starrocks/sql/analyzer/AnalyzerUtils.java [hunk 3]
```diff
@@ -719,6 +726,30 @@ public Void visitView(ViewRelation node, Void context) {
...
+        public JoinPredicateRelationCollector(List<Pair<Expr, Relation[]>> joinPredicates) {
+            super(null);
+            this.joinPredicates = joinPredicates;
+        }
+        @Override
+        public Void visitJoin(JoinRelation node, Void context) {
+            Relation[] relations = new Relation[2];
+            relations[0] = node.getLeft();
+            relations[1] = node.getRight();
+            joinPredicates.add(Pair.create(node.getOnPredicate(), relations));
```
[the remark this answers is not in the document]
Comment (choury) [reply fe/fe-core/src/main/java/com/starrocks/sql/analyzer/AnalyzerUtils.java:798]: Doesn‘t support now

File: fe/fe-core/src/main/java/com/starrocks/catalog/MaterializedView.java [hunk 4]
```diff
@@ -1584,6 +1694,22 @@ public void gsonPostProcess() throws IOException {
                 }
             }
         }
+        partitionExprMaps = Maps.newHashMap();
+        if (serializedPartitionExprMaps != null) {
+            for (Map.Entry<GsonUtils.ExpressionSerializedObject, GsonUtils.ExpressionSerializedObject> entry :
+                    serializedPartitionExprMaps.entrySet()) {
+                if (entry.getKey() != null && entry.getValue() != null) {
+                    partitionExprMaps.put(
+                            SqlParser.parseSqlToExpr(entry.getKey().expressionSql, SqlModeHelper.MODE_DEFAULT),
```
[the remark this answers is not in the document]
Comment (choury) [reply fe/fe-core/src/main/java/com/starrocks/catalog/MaterializedView.java:1601]: refer to `MaterializedViewAnalyzer.checkPartitionColumnWithBaseTable`

File: fe/fe-core/src/main/java/com/starrocks/catalog/MaterializedView.java [hunk 5]
```diff
@@ -1512,6 +1570,47 @@ public Pair<Table, Column> getBaseTableAndPartitionColumn() {
...
+            List<Column> partitionColumns = PartitionUtil.getPartitionColumns(table);
+            if (partitionColumns == null) {
+                continue;
+            }
+            SlotRef slotRef = tableToSlotMap.get(table);
+            if (table.isNativeTableOrMaterializedView()) {
+                if (partitionColumns.size() != 1) {
+                    continue;
+                }
+                Column partitionColumn = partitionColumns.get(0);
```
[the remark this answers is not in the document]
Comment (choury) [reply fe/fe-core/src/main/java/com/starrocks/catalog/MaterializedView.java:1489]: refer to `MaterializedViewAnalyzer.checkPartitionColumnWithBaseTable`

**Review by murphyatwork — APPROVED** [review]

File: fe/fe-core/src/main/java/com/starrocks/catalog/MaterializedView.java [hunk 6]
```diff
@@ -407,6 +407,10 @@ public MvRefreshScheme copy() {
     @SerializedName(value = "maxMVRewriteStaleness")
     private int maxMVRewriteStaleness = 0;
 
+    @SerializedName(value = "partitionExprMaps")
+    private Map<GsonUtils.ExpressionSerializedObject, GsonUtils.ExpressionSerializedObject> serializedPartitionExprMaps;
+    private Map<Expr, SlotRef> partitionExprMaps;
```
Comment (LiShuMing) [inline fe/fe-core/src/main/java/com/starrocks/catalog/MaterializedView.java:412]: What if the MV is upgraded from a lower version and `partitionExprMaps ` is null?  can be compatible with old version?

Comment (choury) [reply fe/fe-core/src/main/java/com/starrocks/catalog/MaterializedView.java:412]: It will be generated from `partitionRefTableExprs` in `gsonPostProcess`

**Review by packy92 — APPROVED** [review]

**Pull request merged**